shapely/tests/test_io.py

```diff
@@ -266,6 +269,55 @@ def test_from_wkb_all_types(geom, use_hex, byte_order):
     assert_geometries_equal(actual, geom)
 
 
+@pytest.mark.parametrize("geom", all_types_with_z)
+@pytest.mark.parametrize("use_hex", [False, True])
+@pytest.mark.parametrize("byte_order", [0, 1])
+def test_from_wkb_all_types_with_z(geom, use_hex, byte_order):
+    if shapely.get_type_id(geom) == shapely.GeometryType.LINEARRING:
+        pytest.skip("Linearrings are not preserved in WKB")
+    wkb = shapely.to_wkb(geom, hex=use_hex, byte_order=byte_order)
+    actual = shapely.from_wkb(wkb)
+    assert_geometries_equal(actual, geom)
+
+
+@pytest.mark.skipif(
+    shapely.geos_version < (3, 12, 0),
+    reason="M coordinates not supported with GEOS < 3.12",
+)
+@pytest.mark.parametrize("geom", all_types_with_m)
+@pytest.mark.parametrize("use_hex", [False, True])
+@pytest.mark.parametrize("byte_order", [0, 1])
+def test_from_wkb_all_types_with_m(geom, use_hex, byte_order):
+    if shapely.get_type_id(geom) == shapely.GeometryType.LINEARRING:
+        pytest.skip("Linearrings are not preserved in WKB")
+    wkb = shapely.to_wkb(geom, hex=use_hex, byte_order=byte_order)
+    actual = shapely.from_wkb(wkb)
+    assert_geometries_equal(actual, geom)
+    assert not actual.has_z
+    # TODO: assert actual.has_m
+
+
+@pytest.mark.skipif(
+    shapely.geos_version < (3, 12, 0),
+    reason="M coordinates not supported with GEOS < 3.12",
+)
+@pytest.mark.parametrize("geom", all_types_with_zm)
+@pytest.mark.parametrize("use_hex", [False, True])
+@pytest.mark.parametrize("byte_order", [0, 1])
+def test_from_wkb_all_types_with_zm(geom, use_hex, byte_order):
+    if shapely.get_type_id(geom) == shapely.GeometryType.LINEARRING:
+        pytest.skip("Linearrings are not preserved in WKB")
+    wkb = shapely.to_wkb(geom, hex=use_hex, byte_order=byte_order)
+    actual = shapely.from_wkb(wkb)
+    assert_geometries_equal(actual, geom)
+    if actual.is_empty:
+        # TODO: Z gets enabled for some reason
```

Comment:
enabled -> disabled?

Reply:
So one example here (with a recent GEOS main for a few weeks ago, not the latest, so things might have changed):

```
In [1]: import shapely

In [2]: p = shapely.from_wkt("POINT ZM EMPTY")

In [3]: p
Out[3]: <POINT Z EMPTY>

In [4]: p.has_z
Out[4]: False

In [5]: shapely.get_coordinate_dimension(p)
Out[5]: 4

In [6]: shapely.to_wkt(p, output_dimension=4)
Out[6]: 'POINT Z EMPTY'

In [7]: shapely.to_wkt(p, output_dimension=3)
Out[7]: 'POINT Z EMPTY'

In [8]: shapely.to_wkt(p, output_dimension=2)
Out[8]: 'POINT Z EMPTY'
```

The overall dimensionality is correct, but `has_z` incorrectly gives False, and the WKT output incorrectly only has Z and not M. 
Also forcing to only show an output dimension up to 2 doesn't drop the "Z".

Reply:
I noticed that even with the current release `shapely.from_wkt("POINT Z EMPTY").has_z is True`, which I consider a bug. I'm considering splitting some of this PR into maintenance fixes for 2.0, as the scope is creeping out of enabling M coordinates.